src/rust/ensogl/src/application/shortcut.rs

```diff
@@ -75,7 +84,22 @@ impl Registry {
 }
 
 impl RegistryModel {
+
+    /// Check the action against internal state and potentially change it. Right now this is used
+    /// to detect DoublePress actions.
+    fn pre_process_action(&self, action_type:ActionType, key_mask:&KeyMask) -> ActionType {
+        let mut detector     =  self.double_press_detector.borrow_mut();
+        let is_double_press  = detector.process_action(action_type,key_mask);
+        if is_double_press {
+            ActionType::DoublePress
+        } else {
+            action_type
+        }
+    }
```

Comment:
I really don't like this design. There was an FRP network that decides whether this is press or release event. The idea is that FRP is the controller (drives the whole logic), and then uses functions to change the model. Now, the FRP is part of the stateful controller, discovers action like press or release, then functions it calls can change the action by adding some more state and control logic. Instead, the discovery whether it was press / release / double press should be all in FRP.